platform-sdk/docs/proposals/consensus-layer/Consensus-Layer.md

```diff
@@ -0,0 +1,669 @@
+birth-round filter. This prevents Alice from being overwhelmed by events.
+
+Further, since events are coming more slowly to Alice, she will also create her own events more slowly.
+
+Further, since events are coming more slowly to Alice, her Event Creator will update more slowly. When the tipset
+algorithm is consulted for an "other parent", it will find after a short time that there are no other parents it can
+select that will advance consensus. This will cause it to stop creating events until enough events from other creators
+have been processed. This natural slowing of event creation provides a virtuous cycle: as each stressed node slows down
+the event creation rate, it starts to create events with more and more transactions within each event and at a slower
+rate. This will lead to fewer overall events, allowing it to do less work. If the node is still not able to keep up,
+eventually it will refuse to accept any additional transactions from users. If enough nodes in the network are stressed,
+then the overall transaction ingestion rate of the network will be reduced, further reducing the amount of work each
+node has to do. Eventually, and equilibrium is reached.
+
+If the rate at which the network is creating events slows, Alice will be able to catch up by retrieving those previously
+dropped events through gossip, and will be able to process them and catch up. Or, in the last extremity where Alice has
+fallen too far behind, Alice will wait for some time and reconnect.
+
+In addition, the Gossip module reports to the Execution layer health information about each node in the network, as
+well as its own health. Health in this case can simply be the highest birth-round received per node in the network. The
+birth-round is a good proxy for how healthy the node is, since slower, unhealthy nodes will have lower birth-rounds
+relative to their peers. Using this information, the Execution layer can throttle transaction ingestion network-wide.
+How this is done will be described in the "dynamic throttles" design documentation.
+
+##### Slow Execution
+
+Under normal circumstances, the Execution layer is always the actual bottleneck. The cost of processing a few hundreds
+of events pales in comparison to the cost of processing tens of thousands of transactions. Execution must therefore
+provide some backpressure on Consensus. In this design, we propose that the Hashgraph module **never** runs the
+consensus algorithm until it is told to do so from the Execution module.
+
+The TCP Sliding Window is a classic technique for controlling backpressure in networking. The receiver controls the rate
+of transmission by signalling to the sender how many bytes can be sent before receiving a new update from the receiver.
+The same technique is used for backpressure in the HTTP2 protocol. We will use the same concept here.
+
+Execution will instruct Consensus each time it needs additional rounds processed. It could indicate any number of
```

Comment:
There are some nice side effects of this design. But here are some ways to reason about it.

First, if we've done this right, we **should** always see that the limit in the system is how long it takes to handle a transaction. Or hash, or sign, or upload a block. All of these things happen in Execution. In addition, Execution is the place with the transaction throttles, and the place where the logic would be to adjust those throttles dynamically based on throughput.

Since Execution will be the bottleneck, whether there is 1 round prepared by Consensus before Execution is ready or 10M rounds makes no difference to Execution. As long as 1 round is ready when Execution needs it, from Execution's perspective, Consensus might as well be infinitely fast.

And if Execution is *not* the bottleneck, then it doesn't really matter to Execution, since it has no work to do anyway. So it might as well block and wait.

What we have here is a classic producer/consumer relationship, where the producer is faster than the consumer. In such a case, we *could* put a buffer between producer and consumer, but since the producer is faster than the consumer, it is just going to fill that buffer up and then it will have to block or stall. So there is no avoiding this, the Hashgraph will stall. The only question is, how much memory do you use for buffering.

Now, just because the Hashgraph stalls, doesn't mean Consensus in generally has stopped. We get our "slop" in the system with the birth-round filtering. We define some number of "future birth-round events" that we accept. So if other nodes are faster than us, then we will have lots of events handy at the time the Hashgraph module is asked to produce another round, and it can produce a new round faster than Execution can finish with the one it is currently working with. So we should be good there.

If a node is not behind, but is ahead of everyone else, then it has to slow down anyway, since it cannot come to consensus all by itself, it needs friends. So if I have a much faster machine than everyone else, it spends time idle. But that's good and expected.